x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java

```diff
@@ -9,21 +9,45 @@
 
 import org.elasticsearch.compute.operator.DriverContext;
 import org.elasticsearch.compute.operator.EvalOperator.ExpressionEvaluator;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.xpack.esql.EsqlClientException;
 import org.elasticsearch.xpack.esql.type.EsqlDataTypes;
 import org.elasticsearch.xpack.ql.expression.Expression;
 import org.elasticsearch.xpack.ql.tree.Source;
 import org.elasticsearch.xpack.ql.type.DataType;
 import org.elasticsearch.xpack.ql.type.DataTypes;
 
+import java.time.Duration;
+import java.time.Period;
 import java.time.temporal.TemporalAmount;
 import java.util.function.Function;
-import java.util.function.Predicate;
 
 import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.DATE_PERIOD;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.TIME_DURATION;
 import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.isDateTimeOrTemporal;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.isTemporalAmount;
 
 abstract class DateTimeArithmeticOperation extends EsqlArithmeticOperation {
 
+    /**
+     * Custom exception to handle e.g. overflows when folding temporal values; we want to set the correct HTTP status (400).
+     */
+    private static class IllegalTemporalValueException extends EsqlClientException {
+        protected IllegalTemporalValueException(String message, Object... args) {
+            super(message, args);
+        }
+
+        @Override
+        public RestStatus status() {
+            return RestStatus.BAD_REQUEST;
+        }
+
+        public static IllegalTemporalValueException fromArithmeticException(Source source, ArithmeticException e) {
+            return new IllegalTemporalValueException("arithmetic exception in expression [{}]: [{}]", source.text(), e.getMessage());
+        }
+    }
```

Comment:
An exception has 'cognitive' value - it should be generic enough that enough places encounter the same condition that an exception makes sense to be created.

This is rarely the case (which is why *QL in general has only a couple of exceptions) -  just return an EsqlClientException with the correct message.

Reply:
I tried just using EsqlClientException before I arrived at this solution; the problem with `EsqlClientException` is that it is abstract, so I can only create an anonymous class that inherits from it.

This works, but the user facing error message that it creates when an overflow is triggered during folding is not so great:
```
{"type":"",
  "reason":"arithmetic exception in expression [2147483647 day + 1 day]: [integer overflow]"}
```
Notice the empty type string. Since the point of this exception is providing a nicer user-facing error, I don't think an anonymous class is a good solution.

With the current solution, the error message looks like this instead:
```
{"type":"illegal_temporal_value_exception",
  "reason":"arithmetic exception in expression [9223372036854775807 second + 1000 millisecond]: [long overflow]"}
```

I'll add this to the javadoc of the class to make it clear.

(I also contemplated either adding a more generic named exception, like `FoldingException` or so - but the case that it's used in only 2 places in just 1 class made that feel very awkward. The other option would be to make EsqlClientException just not abstract, but that again seemed like being against EsqlClientException's intended purpose. If you'd like me to go with any of these solutions, instead, just let me know and I'll create a follow-up PR!

There's also a solid chance that there may be more situations where a user-caused error will happen during folding time, at which point we can also factor out `IllegalTemporalValueException` into a more generic exception class.)

Reply:
As it turns out, I need an exception to communicate arithmetic exceptions when folding `Neg` as well. So, the best is probably to factor this out into a dedicated `DateMathException`.

I'll make a follow-up PR to handle `Neg` and will do that there.